File: 75 days Challenge Leetcode/PrefixSum/Day-18/Solution.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    int largestAltitude(vector<int>& gain) {
        int n = gain.size();
        vector<int> prefix_sum(n + 1); // +1 to include the initial altitude
        prefix_sum[0] = 0; // Starting altitude
        for (int i = 1; i <= n; i++) 
            prefix_sum[i] = prefix_sum[i - 1] + gain[i - 1]; // Add each gain to previous altitude
        return *max_element(prefix_sum.begin(), prefix_sum.end()); // Find the max altitude
    }
};

File: 75 days Challenge Leetcode/PrefixSum/Day-18/Solution.java
import java.util.*;

class Solution {
    public int largestAltitude(int[] gain) {
        int n = gain.length;
        int[] prefixSum = new int[n + 1]; // +1 to include the initial altitude
        prefixSum[0] = 0; // Starting altitude
        
        // Calculate prefix sum for each altitude after each gain
        for (int i = 1; i <= n; i++) {
            prefixSum[i] = prefixSum[i - 1] + gain[i - 1]; // Add gain[i-1] to previous altitude
        }

        // Find the maximum altitude from the prefixSum array
        int maxAltitude = Arrays.stream(prefixSum).max().getAsInt(); // max() returns an OptionalInt, so getAsInt() retrieves it
        return maxAltitude;
    }
}
